accepted/2024/tensors/README.md

```diff
@@ -0,0 +1,424 @@
+    public int Rank { get; }
+    public ReadOnlySpan<nint> Strides { get; }
+
+    public ref readonly T this[params ReadOnlySpan<nint> indices] { get; }
+
+    public void Clear();
+    public void CopyTo(TensorSpan<T> destination);
+    public void Fill(T value);
+    public Enumerator GetEnumerator();
+    public ref readonly T GetPinnableReference();
+    public ReadOnlyTensorSpan<T> Slice(params ReadOnlySpan<Range<nint>> ranges);
+    public bool TryCopyTo(TensorSpan<T> destination);
+
+    public ref struct Enumerator
+    {
+        public ref readonly T Current { get; }
+
+        public bool MoveNext ();
+    }
+}
+```
+
+### Creation and Interop
+
+Interop with other tensor libraries is a pretty crucial factor and while the `TensorSpan<T>` and `ITensor<TSelf, T>` types do cover a large majority of cases, there also exists some more general considerations around unsafely wrapping or getting other access to the data.
+
+For external consumers getting access to the underlying `Tensor<T>`, they can use the combination of `bool IsPinned { get; }` and `ref T GetPinnableReference` alongside `System.Runtime.CompilerServices.Unsafe` to get the underlying data and handle it correctly. A `Tensor<T>` then has a few creation APIs exposed allowing some minimal customization of how the data is allocated. For creating a `Tensor<T>` over other data, some explict unsafe creation APIs are then exposed.
+
+Given the below, the first two APIs allow creating a `Tensor<T>` over runtime allocated memory. If `mustPin` is true then the data is required to be pinned and `bool IsPinned { get; }` will return `true`. This does not necessarily mean the memory has been allocated using native memory and where the memory is allocated remains at the discression of the runtime. There exists a version that only takes `lengths` and assumes there are no gaps between the dimensions. There then exists a separate API which takes the strides in order to support other types of contiguous memory.
+
+There are then separate APIs which directly take a `T*` and which allow creation of a `Tensor<T>` over existing memory. In such a scenario, the tensor assumes no ownership of the data and it remains the responsibiltiy of the caller to free the data when they are done with it. It may be desirable to consider a separate set of APIs that take some `IMemoryAllocator` like interface or even a `Free` delegate where the user can customize the free behavior, but that is not considered in scope for this design. It may be desirable to only
+
+```csharp
+public static partial class Tensor
+{
+    public static Tensor<T> Create(bool mustPin, ReadOnlySpan<nint> lengths);
+    public static Tensor<T> Create(bool mustPin, ReadOnlySpan<nint> lengths, ReadOnlySpan<nint> strides);
+
+    public static Tensor<T> Create(T* address, ReadOnlySpan<nint> lengths);
+    public static Tensor<T> Create(T* address, ReadOnlySpan<nint> lengths, ReadOnlySpan<nint> strides);
```

Comment:
How would we expect existing tensor types to integrate? e.g. TorchSharp's `Tensor<T>`, would the idea be that it would implement `ITensor<Tensor<T>, T>`, or otherwise expose such an implementation?

Reply:
Yes. Those tensor types would choose one or more of:
* Implement `ITensor<TSelf, T>`
* Provide conversions to/from `TensorSpan<T>`
* Provide access to the underlying `T*` so that users can unsafely create a `Tensor<T>` or `TensorSpan<T>` over it

What they get out of it varies between the options, but it gives flexibility for how integrated they want to be with whatever the BCL provides here. The minimal interface should make it very easy for opt-in, even if explicit implementation only.